From 3e6d7cbd5d6c52633f2976cdac3766a1f4552aeb Mon Sep 17 00:00:00 2001
From: "Patil, Rachna" <rachna@ti.com>
Date: Tue, 7 Aug 2012 11:24:42 +0800
Subject: [PATCH 519/609] input: TSC: Shutdown ADC, if TSC wake is not
 required

Original commit: ff509f2d0efb8cc1197b0c599a9ac9483e2d25f5

  From git://arago-project.org/git/projects/linux-am33x.git
  And ti-sdk-am335x-evm-05.05.00.00-Linux-x86-Install image

Currently ADC draws additional power.
This can be brought down by shutting down ADC, if
touchscreen wake is not required.

Signed-off-by: Patil, Rachna <rachna@ti.com>
---
 drivers/input/touchscreen/ti_tscadc.c |   21 +++++++++++++++------
 1 files changed, 15 insertions(+), 6 deletions(-)

diff --git a/drivers/input/touchscreen/ti_tscadc.c b/drivers/input/touchscreen/ti_tscadc.c
index 6689892..abf879c 100644
--- a/drivers/input/touchscreen/ti_tscadc.c
+++ b/drivers/input/touchscreen/ti_tscadc.c
@@ -86,6 +86,7 @@
 #define CNTRLREG_5WIRE		(1 << 6)
 #define CNTRLREG_8WIRE		(3 << 5)
 #define CNTRLREG_TSCENB		(1 << 7)
+#define CNTRLREG_POWERDOWN	BIT(4)
 #define ADCFSM_STEPID		0x10
 
 #define SEQ_SETTLE		275
@@ -480,14 +481,16 @@ static int tscadc_suspend(struct platform_device *pdev, pm_message_t state)
 		idle = tscadc_readl(ts_dev, REG_IRQENABLE);
 		tscadc_writel(ts_dev, REG_IRQENABLE,
 				(idle | IRQENB_HW_PEN));
+		tscadc_writel(ts_dev, REG_SE, 0x00);
 		tscadc_writel(ts_dev, REG_IRQWAKEUP, IRQWKUP_ENB);
-	}
+	} else {
 
-	/* module disable */
-	idle = 0;
-	idle = tscadc_readl(ts_dev, REG_CTRL);
-	idle &= ~(CNTRLREG_TSCSSENB);
-	tscadc_writel(ts_dev, REG_CTRL, idle);
+		/* module disable */
+		idle = tscadc_readl(ts_dev, REG_CTRL);
+		idle &= ~(CNTRLREG_TSCSSENB);
+		tscadc_writel(ts_dev, REG_CTRL, (idle |
+				CNTRLREG_POWERDOWN));
+	}
 
 	pm_runtime_put_sync(&pdev->dev);
 
@@ -509,6 +512,12 @@ static int tscadc_resume(struct platform_device *pdev)
 
 	/* context restore */
 	tscadc_writel(ts_dev, REG_CTRL, ts_dev->ctrl);
+
+	/* Make sure ADC is powered up */
+	restore = tscadc_readl(ts_dev, REG_CTRL);
+	restore &= ~(CNTRLREG_POWERDOWN);
+	tscadc_writel(ts_dev, REG_CTRL, restore);
+
 	tsc_idle_config(ts_dev);
 	tsc_step_config(ts_dev);
 	tscadc_writel(ts_dev, REG_FIFO1THR, 6);
-- 
1.7.5.4

